File: nodecore-spv/src/main/java/veriblock/wallet/Ledger.java
// VeriBlock Blockchain Project
// Copyright 2017-2018 VeriBlock, Inc
// Copyright 2018-2020 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package veriblock.wallet;

import org.veriblock.sdk.models.Coin;
import veriblock.model.Output;
import veriblock.model.StandardTransaction;
import veriblock.model.TransactionMeta;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Ledger {
    private Map<String, AddressLedger> entries = new HashMap<>();

    public void add(String address) {
        entries.putIfAbsent(address,
                new AddressLedger(address, Coin.ZERO, 0));
    }

    public AddressLedger get(String address) {
        return entries.get(address);
    }

    public Collection<AddressLedger> list() {
        return entries.values();
    }

    public void load(List<AddressLedger> toLoad) {
        entries.clear();
        for (AddressLedger addressLedger : toLoad) {
            add(addressLedger);
        }
    }

    public void record(StandardTransaction tx) {
        List<LedgerEntry> ledgerEntries = createLedgerEntriesFrom(tx);
        ledgerEntries.forEach(entry -> get(entry.getAddress()).add(entry));
    }

    private void add(AddressLedger addressLedger) {
        entries.putIfAbsent(addressLedger.getAddress(), addressLedger);
    }

    private List<LedgerEntry> createLedgerEntriesFrom(StandardTransaction tx) {
        List<LedgerEntry> ledgerEntries = new ArrayList<>();

        LedgerEntry.Status status = (tx.getTransactionMeta().getState() == TransactionMeta.MetaState.CONFIRMED) ? LedgerEntry.Status.CONFIRMED : LedgerEntry.Status.PENDING;
        if (entries.containsKey(tx.getInputAddress().get())) {
            ledgerEntries.add(new LedgerEntry(tx.getInputAddress().get(),
                    tx.getTxId(),
                    tx.getInputAmount(),
                    Coin.ZERO,
                    tx.getSignatureIndex(),
                    0,
                    status));
        }

        for (int i = 0; i < tx.getOutputs().size(); i++) {
            Output o = tx.getOutputs().get(i);
            if (entries.containsKey(o.getAddress().get())) {
                ledgerEntries.add(new LedgerEntry(o.getAddress().get(),
                        tx.getTxId(),
                        Coin.ZERO,
                        o.getAmount(),
                        -1,
                        i,
                        status));
            }
        }

        return ledgerEntries;
    }
}

File: nodecore-spv/src/main/java/org/veriblock/spv/wallet/Ledger.kt
// VeriBlock Blockchain Project
// Copyright 2017-2018 VeriBlock, Inc
// Copyright 2018-2021 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package org.veriblock.spv.wallet

import org.veriblock.sdk.models.Coin
import org.veriblock.sdk.models.VeriBlockTransaction
import org.veriblock.spv.model.StandardTransaction
import org.veriblock.spv.model.TransactionMeta
import java.util.ArrayList
import java.util.HashMap

class Ledger {
    private val entries: MutableMap<String, AddressLedger> = HashMap()

    fun add(address: String) {
        entries.putIfAbsent(
            address,
            AddressLedger(address, Coin.ZERO, 0)
        )
    }

    operator fun get(address: String): AddressLedger {
        return entries.getValue(address)
    }

    fun list(): Collection<AddressLedger> {
        return entries.values
    }

    fun load(toLoad: List<AddressLedger>) {
        entries.clear()
        for (addressLedger in toLoad) {
            add(addressLedger)
        }
    }

    fun record(tx: StandardTransaction) {
        val ledgerEntries = createLedgerEntriesFrom(tx)
        ledgerEntries.forEach {
            get(it.address).add(it)
        }
    }

    private fun add(addressLedger: AddressLedger) {
        entries.putIfAbsent(addressLedger.address, addressLedger)
    }

    private fun createLedgerEntriesFrom(tx: StandardTransaction): List<LedgerEntry> {
        val ledgerEntries = ArrayList<LedgerEntry>()
        val status = if (tx.meta.state === TransactionMeta.MetaState.CONFIRMED) {
            LedgerEntry.Status.CONFIRMED
        } else {
            LedgerEntry.Status.PENDING
        }
        if (entries.containsKey(tx.tx.sourceAddress.address)) {
            ledgerEntries.add(
                LedgerEntry(tx.tx.sourceAddress.address, tx.tx.id, tx.tx.sourceAmount, Coin.ZERO, tx.tx.signatureIndex, 0, status)
            )
        }
        for (i in tx.tx.outputs.indices) {
            val o = tx.tx.outputs[i]
            if (entries.containsKey(o.address.address)) {
                ledgerEntries.add(
                    LedgerEntry(o.address.address, tx.tx.id, Coin.ZERO, o.amount, -1, i, status)
                )
            }
        }
        return ledgerEntries
    }
}
